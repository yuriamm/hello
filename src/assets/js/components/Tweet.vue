<template>
  <div>
    <NavBar :isLoggedIn="isLoggedIn" />
    <div v-if="isLoggedIn">
      <div class="alert alert-danger" v-if="error">Error!</div>
      <div class="alert alert-success" v-if="success">Success!</div>
      <form @submit.prevent="tweet">
        <input v-model="post" type="text" required />

        <button type="submit">Tweet</button>
      </form>
    </div>
    <div v-for="tweet in tweets" :key="tweet.id">
      <span>{{ tweet.tweet }}</span>
      <div v-if="isLoggedIn">
        <button
          v-if="tweet.user_id === parseInt(user_id)"
          @click="remove(tweet.id)"
        >
          Delete
        </button>
        <button v-if="isFavorited(tweet)" @click="unfavorite(tweet.id)">
          Unfavorite
        </button>
        <button v-else @click="favorite(tweet.id)">
          Favorite
        </button>
      </div>
    </div>
  </div>
</template>

<script>
import request from "../api/request";
import NavBar from "./NavBar";
export default {
  name: "tweet",
  components: {
    NavBar
  },
  data() {
    return {
      tweets: [],
      post: "",
      user_id: 0,
      error: false,
      success: false,
      isLoggedIn: false
    };
  },
  async created() {
    const user_id = sessionStorage.getItem("user_id");
    if (user_id) {
      this.isLoggedIn = true;
      this.user_id = user_id;
    }
  },
  async mounted() {
    const response = await request("/api/home", {}, "get");
    this.tweets = response.data.tweets;
  },
  methods: {
    isFavorited(tweet) {
      return tweet.users_favorited.includes(parseInt(this.user_id));
    },
    async tweet() {
      const response = await request("/api/home", { tweet: this.post }, "post");
      if (response.data.message === "success") {
        this.tweets = [response.data, ...this.tweets];
        window.location.href = "/home";
        this.success = true;
      }
      if (response.data.message === "error") {
        this.error = true;
      }
    },
    async remove(id) {
      const response = await request(`/api/home/${id}`, {}, "delete");
      if (response.data.message === "success") {
        window.location.href = "/home";
        this.success = true;
      }
      if (response.data.message === "error") {
        this.error = true;
      }
    },
    async favorite(id) {
      const response = await request(`/api/favorite`, { id: id }, "post");
      if (response.data.message === "success") {
        window.location.href = "/home";
        this.success = true;
      }
      if (response.data.message === "error") {
        this.error = true;
      }
    },
    async unfavorite(id) {
      const response = await request(`/api/favorite/${id}`, {}, "delete");
      if (response.data.message === "success") {
        window.location.href = "/home";
        this.success = true;
      }
      if (response.data.message === "error") {
        this.error = true;
      }
    }
  }
};
</script>
