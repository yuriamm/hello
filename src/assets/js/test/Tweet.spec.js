import { shallowMount } from "@vue/test-utils";
import Tweet from "../components/Tweet.vue";
jest.mock("../api/request.js");

global.window = Object.create(window);
const url = "http://localhost:4000";
Object.defineProperty(window, "location", {
  value: {
    href: url,
  },
});

const wrapper = shallowMount(Tweet);

describe("Tweet.vue", () => {
  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders correctly when logged in", async () => {
    await wrapper.setData({
      isLoggedIn: true,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it("redirects to correct path when tweet", async () => {
    await wrapper.setData({
      post: "post",
      isLoggedIn: true,
    });
    const button = wrapper.findAll("button").at(0);
    expect(button.text()).toEqual("Tweet");

    wrapper.find("form").trigger("submit.prevent");
    wrapper.vm.$nextTick(() => {
      expect(window.location.href).toEqual("/home");
      done();
    });
  });

  it("redirects to correct path when delete", async () => {
    await wrapper.setData({
      tweets: [
        {
          tweet: "tweet",
          id: 1,
          user_id: 0,
          users_favorited: [],
        },
      ],
      user_id: "0",
      isLoggedIn: true,
    });
    const button = wrapper.findAll("button").at(1);
    expect(button.text()).toEqual("Delete");

    button.trigger("click");
    wrapper.vm.$nextTick(() => {
      expect(window.location.href).toEqual("/home");
      done();
    });
  });

  it("redirects to correct path when unfavorite", async () => {
    await wrapper.setData({
      tweets: [
        {
          tweet: "tweet",
          id: 1,
          user_id: 1,
          users_favorited: [0],
        },
      ],
      user_id: "0",
      isLoggedIn: true,
    });
    const button = wrapper.findAll("button").at(1);
    expect(button.text()).toEqual("Unfavorite");

    button.trigger("click");
    wrapper.vm.$nextTick(() => {
      expect(window.location.href).toEqual("/home");
      done();
    });
  });

  it("redirects to correct path when favorite", async () => {
    await wrapper.setData({
      tweets: [
        {
          tweet: "tweet",
          id: 1,
          user_id: 1,
          users_favorited: [1],
        },
      ],
      user_id: "0",
      isLoggedIn: true,
    });
    const button = wrapper.findAll("button").at(1);
    expect(button.text()).toEqual("Favorite");

    button.trigger("click");
    wrapper.vm.$nextTick(() => {
      expect(window.location.href).toEqual("/home");
      done();
    });
  });
});
