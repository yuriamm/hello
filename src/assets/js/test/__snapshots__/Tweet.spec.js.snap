// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tweet.vue renders correctly 1`] = `
<div>
  <navbar-stub />
   
  <!---->
   
  <div>
    <span>
      
    </span>
     
    <span>
      favorited: 0
    </span>
     
    <!---->
  </div>
</div>
`;

exports[`Tweet.vue renders correctly when logged in 1`] = `
<div>
  <navbar-stub
    isloggedin="true"
  />
   
  <div>
    <!---->
     
    <!---->
     
    <form>
      <input
        required="required"
        type="text"
      />
       
      <button
        type="submit"
      >
        Tweet
      </button>
    </form>
  </div>
   
  <div>
    <span>
      
    </span>
     
    <span>
      favorited: 0
    </span>
     
    <div>
      <button>
        
        Delete
      
      </button>
       
      <button>
        
        Favorite
      
      </button>
    </div>
  </div>
</div>
`;
